File: days/day_04/main.js
const fs = require('fs');
const readline = require('readline');

async function loadInput() {
  const rl = readline.createInterface({
    input: fs.createReadStream('input.txt')
  });

  const lines = [];
  for await (const line of rl) {
    lines.push(line);
  }
  return lines
}

function stringToPassport(s) {
  const passport = {}
  const items = s.trim(' ').split(' ').map(x => x.split(':'))
  
  for (const item of items) {
    passport[item[0]] = item[1]
  }

  return passport
}

function allFieldsPresent(passport) {
  const fields = ['byr','iyr','eyr','hgt','hcl','ecl','pid'];
  for (const field of fields){
    if (passport[field] == undefined) {
      return false;
    }
  }
  return true;
}

function validate(passport) {
  if (!allFieldsPresent(passport)) {
    return false
  }

  byr = passport['byr']
  if (!/^(19[2-9]\d|200[0-2])$/g.test(byr)) {
    return false
  }

  iyr = passport['iyr']
  if (!/^(201\d|2020)$/.test(iyr)) {
    return false
  }

  eyr = passport['eyr']
  if (!/^(202\d|2030)$/.test(eyr)) {
    return false
  }

  hgt = passport['hgt']
  if (!/^(1[5-8]\d|19[0-3])cm$/.test(hgt) && !/^(59|6\d|7[0-6])in$/.test(hgt)) {
    return false
  }

  hcl = passport['hcl']
  if (!/^#[0-9a-f]{6}$/.test(hcl)) {
    return false
  }

  ecl = passport['ecl']
  if (!['amb','blu','brn','gry','grn','hzl','oth'].includes(ecl)) {
    return false
  }

  pid = passport['pid']
  if (!/^\d{9}$/.test(pid)) {
    return false
  }

  return true
}

function partOne(passports) {
  let total_valid = 0
  for (const passport of passports){
    total_valid += allFieldsPresent(passport);
  }
  return total_valid
}

function partTwo(passports) {
 let total_valid = 0
  for (const passport of passports){
    total_valid += validate(passport);
  }
  return total_valid
}

async function main() {
  const start_time = process.hrtime() 
  
  const lines = await loadInput();

  const passports = []

  let s = ''
  for (const line of lines) {
    if (line == ''){
      passports.push(stringToPassport(s))
      s = ''
    }
    else {
      s += line + ' '
    }
  }
  passports.push(stringToPassport(s))

  const part_one_ans = partOne(passports)
  const part_two_ans = partTwo(passports)
  
  const diff = process.hrtime(start_time)
  const total_time = (diff[0] + diff[1]/1e9).toFixed(3)
  
  console.log(`Day  1 (${total_time}s)`)
  console.log(`  Part 1: ${part_one_ans}`);
  console.log(`  Part 2: ${part_two_ans}`);
}

if (require.main === module) {
  main();
}

File: days/day_04/main.ml
open Core

let file = "input.txt"

let string_to_passport s =
	let items = s
		|> String.strip 
		|> String.split ~on:' '
		|> List.map ~f:(fun item ->
			match String.split item ~on:':' with
			| a :: b :: [] -> (a, b)
			| _ -> failwith "nope"
		)
	in
	Map.of_alist_exn (module String) items
;;

let all_fields_present passport =
  let rec check passport fields =
  	match fields with
  	| field :: tl ->
  		if (Map.mem passport field) then check passport tl
  		else false
  	| [] -> true
  in
  check passport ["byr";"iyr";"eyr";"hgt";"hcl";"ecl";"pid"]
;;

let is_valid passport = 
	if not (all_fields_present passport) then false
	else if 
		let byr = Map.find_exn passport "byr" in
		not (Str.string_match (Str.regexp "^\\(19[2-9][0-9]\\|200[0-2]\\)$") byr 0)
		then false
	else if 
		let iyr = Map.find_exn passport "iyr" in
		not (Str.string_match (Str.regexp "^\\(201[0-9]\\|2020\\)$") iyr 0)
		then false
	else if 
		let eyr = Map.find_exn passport "eyr" in
		not (Str.string_match (Str.regexp "^\\(202[0-9]\\|2030\\)$") eyr 0)
		then false
	else if 
		let hgt = Map.find_exn passport "hgt" in
		not (Str.string_match (Str.regexp "^\\(1[5-8][0-9]\\|19[0-3]\\)cm$") hgt 0)
		&& not  (Str.string_match (Str.regexp "^\\(59\\|6[0-9]\\|7[0-6]\\)in$") hgt 0)
		then false
	else if 
		let hcl = Map.find_exn passport "hcl" in
		not (Str.string_match (Str.regexp "^#[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]$") hcl 0)
		then false
	else if 
		let ecl = Map.find_exn passport "ecl" in
		not (List.mem ["amb"; "blu"; "brn"; "gry"; "grn"; "hzl"; "oth"] ecl ~equal:String.(=))
		then false
	else if 
		let pid = Map.find_exn passport "pid" in
		not (Str.string_match (Str.regexp "^[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]$") pid 0)
		then false
	else
		true
;;

let part_one passports =
	List.fold passports ~init:0 ~f:(fun total passport ->
		if (all_fields_present passport) then (total+1)
		else total
	)
;;

let part_two passports =
	List.fold passports ~init:0 ~f:(fun total passport ->
		if (is_valid passport) then (total+1)
		else total
	)
;;

let () = 
	let start_time = Unix.gettimeofday () in
	let input = In_channel.read_lines file in
	let passports =
		let rec parse_input lines passports s = 
			match lines with
			| line :: tl -> 
				if String.(=) line "" 
				then parse_input 
					tl
					(passports @ [(string_to_passport s)])
					""
				else
					parse_input tl passports (String.concat [s; line; " "])
			| [] -> passports @ [(string_to_passport s)]
		in
		parse_input input [] ""
	in
	let part_one_ans = part_one passports in
	let part_two_ans = part_two passports in
	printf "Day  4 (%.3fs)\n" ((Unix.gettimeofday ()) -. start_time);
	printf "  Part 1: %d\n" part_one_ans;
	printf "  Part 2: %d\n" part_two_ans
;;
